acnportal/acnsim/base.py

```diff
@@ -0,0 +1,473 @@
+            UserWarning: If any of the unhandled attributes are not
+                JSON serializable.
+
+        """
+        context_dict, = none_to_empty_dict(context_dict)
+        obj_id = f'{id(self)}'
+
+        # Check if this object has already been converted, and return
+        # the appropriate dict if this is the case.
+        if obj_id in context_dict:
+            return {'id': obj_id, 'context_dict': context_dict}, context_dict
+
+        obj_type = f'{self.__module__}.{self.__class__.__name__}'
+
+        # Get a dictionary of the attributes of this object using the
+        # object's to_dict method.
+        attribute_dict, context_dict = self.to_dict(context_dict)
+
+        # Check that all attributes have been serialized.
+        # Warn if some attributes weren't serialized.
+        if attribute_dict.keys() != self.__dict__.keys():
+            unserialized_keys = \
+                set(self.__dict__.keys()) - set(attribute_dict.keys())
+            warnings.warn(
+                f"Attributes {unserialized_keys} present in object of "
+                f"type {obj_type} but not handled by object's to_dict "
+                f"method. Serialized object may not load correctly. "
+                f"Write a to_dict method and re-dump, or write an "
+                f"appropriate from_dict method to accurately load.",
+                UserWarning
+            )
+            for key in unserialized_keys:
+                unserialized_attr = self.__dict__[key]
+                # Try calling the attr's to_registry method.
+                try:
+                    registry, context_dict = unserialized_attr.to_registry(
+                        context_dict=context_dict)
+                    attribute_dict[key] = registry['id']
+                    continue
+                except AttributeError:
+                    pass
+                # Try dumping the object using the native JSON
+                # serializer.
+                try:
+                    json.dumps(unserialized_attr)
+                except TypeError:
+                    warnings.warn(
+                        f"Attribute {key} could not be serialized. "
+                        f"Dumping the attribute's repr() "
+                        f"representation. This attribute will not be "
+                        f"fully loaded.",
+                        UserWarning
+                    )
+                    attribute_dict[key] = repr(unserialized_attr)
+                else:
+                    attribute_dict[key] = unserialized_attr
+
+        obj_dict = {'class': obj_type, 'attributes': attribute_dict}
+        context_dict[obj_id] = obj_dict
+
+        # Check versions of acnportal and certain dependencies.
+        acnportal_version = pkg_resources.require('acnportal')[0].version
+        dependency_versions = {'numpy': numpy.__version__,
+                               'pandas': pandas.__version__}
```

Comment:
Here, dependency versions should not be checked if this is a recursive call. Check that this is the case.